server/observer.go

```diff
@@ -0,0 +1,199 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/grpc"
+)
+
+// UnaryObserver provides an interface for observing unary requests and responses.
+type UnaryObserver interface {
+	GetName() string
+	ObserveUnary(
+		ctx context.Context,
+		req interface{},
+		resp interface{},
+		info *grpc.UnaryServerInfo,
+		err error)
+}
+
+// StreamRequestObserver provides an interface for observing streaming requests.
+type StreamRequestObserver interface {
+	GetName() string
+	ObserveStreamRequest(
+		ctx context.Context,
+		req interface{},
+		info *grpc.StreamServerInfo,
+		err error)
+}
+
+// StreamResponseObserver provides an interface for observing streaming responses.
+type StreamResponseObserver interface {
+	GetName() string
+	ObserveStreamResponse(
+		ctx context.Context,
+		resp interface{},
+		info *grpc.StreamServerInfo,
+		err error)
+}
+
+// GrpcObserverRegistry is a registry of observers. These observers are hooked into the
+// grpc interceptors that are provided by this interface.
+type GrpcObserverRegistry interface {
+	// UnaryInterceptor implements the grpc.UnaryServerInterceptor type to allow the
+	// registry to hook into unary grpc methods.
+	UnaryInterceptor(
+		context.Context,
+		interface{},
+		*grpc.UnaryServerInfo,
+		grpc.UnaryHandler) (interface{}, error)
+	// StreamInterceptor implements the grpc.StreamServerInterceptor type to allow the
+	// registry to hook into streaming grpc methods.
+	StreamInterceptor(
+		interface{},
+		grpc.ServerStream,
+		*grpc.StreamServerInfo,
+		grpc.StreamHandler) error
+	RegisterUnaryObserver(UnaryObserver)
+	RegisterStreamRequestObserver(StreamRequestObserver)
+	RegisterStreamResponseObserver(StreamResponseObserver)
+}
+
+// ShowcaseObserverRegistry returns the showcase specific observer registry.
+func ShowcaseObserverRegistry() GrpcObserverRegistry {
+	return &showcaseObserverRegistry{
+		uObservers:     map[string]UnaryObserver{},
+		sReqObservers:  map[string]StreamRequestObserver{},
+		sRespObservers: map[string]StreamResponseObserver{},
+	}
+}
+
+// showcaseObserverRegistry is an implementation of the ObserverRegistry. This registry
+// automatically handles DeleteTest requests and deletes the appropriate observers
+// for that request.
+type showcaseObserverRegistry struct {
+	mu             sync.Mutex
+	uObservers     map[string]UnaryObserver
+	sReqObservers  map[string]StreamRequestObserver
+	sRespObservers map[string]StreamResponseObserver
+}
+
+func (r *showcaseObserverRegistry) UnaryInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler) (interface{}, error) {
+	resp, err := handler(ctx, req)
+
+	if info.FullMethod == "/google.showcase.v1alpha3.Testing/DeleteTest" {
+		err = r.deleteTestHandler(req)
```

Comment:
Out of curiosity, is it possible that a delete call beats a unary/stream call using the same observer, grabs the mutex, and deletes the observer? Or are the tests structured such that the unary/stream calls are run in sequence and a delete call will only be complete once all of the related unary tests are complete?

Reply:
Since all the observers are behind the same lock and all access to the observers will lock the mutex. Look down at the impl of the deleteTestHandler, it locks if it is going to attempt to delete the test. Maybe I'm missing something?

Reply:
nvm i talked myself out of confusion as I drafted a response. Thanks.